// @vscode/chat-lib/dist/src/_internal/platform/inlineEdits/common/inlineEditsModelService.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.NullUndesiredModelsManager = exports.IUndesiredModelsManager = exports.IInlineEditsModelService = void 0;
const services_1 = require("../../../util/common/services");
exports.IInlineEditsModelService = (0, services_1.createServiceIdentifier)('IInlineEditsModelService');
exports.IUndesiredModelsManager = (0, services_1.createServiceIdentifier)('IUndesiredModelsManager');
class NullUndesiredModelsManager {
    isUndesiredModelId(_modelId) {
        return false;
    }
    addUndesiredModelId(_modelId) {
        return Promise.resolve();
    }
    removeUndesiredModelId(_modelId) {
        return Promise.resolve();
    }
}
exports.NullUndesiredModelsManager = NullUndesiredModelsManager;
