// @vscode/chat-lib/dist/src/_internal/extension/completions-core/vscode-node/lib/src/auth/copilotTokenManager.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CopilotTokenManagerImpl = exports.ICompletionsCopilotTokenManager = exports.CopilotToken = void 0;
const authentication_1 = require("../../../../../../platform/authentication/common/authentication");
const services_1 = require("../../../../../../util/common/services");
const async_1 = require("../../../../../../util/vs/base/common/async");
const lifecycle_1 = require("../../../../../../util/vs/base/common/lifecycle");
var copilotToken_1 = require("../../../../../../platform/authentication/common/copilotToken");
Object.defineProperty(exports, "CopilotToken", { enumerable: true, get: function () { return copilotToken_1.CopilotToken; } });
exports.ICompletionsCopilotTokenManager = (0, services_1.createServiceIdentifier)('ICompletionsCopilotTokenManager');
let CopilotTokenManagerImpl = class CopilotTokenManagerImpl extends lifecycle_1.Disposable {
    get token() {
        void this.tokenRefetcher.trigger(() => this.updateCachedToken());
        return this._token;
    }
    constructor(primed = false, authenticationService) {
        super();
        this.primed = primed;
        this.authenticationService = authenticationService;
        this.tokenRefetcher = new async_1.ThrottledDelayer(5_000);
        this.updateCachedToken();
        this._register(this.authenticationService.onDidAuthenticationChange(() => this.updateCachedToken()));
    }
    /**
     * Ensure we have a token and that the `StatusReporter` is up to date.
     */
    primeToken() {
        try {
            return this.getToken().then(() => true, () => false);
        }
        catch (e) {
            return Promise.resolve(false);
        }
    }
    async getToken() {
        return this.updateCachedToken();
    }
    async updateCachedToken() {
        this._token = await this.authenticationService.getCopilotToken();
        return this._token;
    }
    resetToken(httpError) {
        this.authenticationService.resetCopilotToken();
    }
    getLastToken() {
        return this.authenticationService.copilotToken;
    }
};
exports.CopilotTokenManagerImpl = CopilotTokenManagerImpl;
exports.CopilotTokenManagerImpl = CopilotTokenManagerImpl = __decorate([
    __param(1, authentication_1.IAuthenticationService)
], CopilotTokenManagerImpl);
