// @vscode/chat-lib/dist/src/_internal/platform/inlineEdits/common/dataTypes/inlineEditsModelsTypes.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.WireTypes = void 0;
const validator_1 = require("../../../configuration/common/validator");
var WireTypes;
(function (WireTypes) {
    let Capabilities;
    (function (Capabilities) {
        function is(obj) {
            return !!obj && typeof obj === 'object' &&
                typeof obj.promptStrategy === 'string';
        }
        Capabilities.is = is;
        Capabilities.validator = (0, validator_1.vObj)({
            promptStrategy: (0, validator_1.vString)(),
        });
    })(Capabilities = WireTypes.Capabilities || (WireTypes.Capabilities = {}));
    let Model;
    (function (Model) {
        Model.validator = (0, validator_1.vObj)({
            serviceType: (0, validator_1.vString)(),
            name: (0, validator_1.vString)(),
            provider: (0, validator_1.vString)(),
            capabilities: Capabilities.validator,
        });
        function is(obj) {
            return !!obj && typeof obj === 'object' &&
                typeof obj.serviceType === 'string' &&
                typeof obj.name === 'string' &&
                typeof obj.provider === 'string' &&
                Capabilities.is(obj.capabilities);
        }
        Model.is = is;
    })(Model = WireTypes.Model || (WireTypes.Model = {}));
    let ModelList;
    (function (ModelList) {
        ModelList.validator = (0, validator_1.vObj)({
            models: (0, validator_1.vArray)(Model.validator),
        });
        function is(obj) {
            return !!obj && typeof obj === 'object' && Array.isArray(obj.models) && obj.models.every(Model.is);
        }
        ModelList.is = is;
    })(ModelList = WireTypes.ModelList || (WireTypes.ModelList = {}));
})(WireTypes || (exports.WireTypes = WireTypes = {}));
