// @vscode/chat-lib/dist/src/_internal/platform/inlineEdits/common/dataTypes/jointCompletionsProviderOptions.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.JointCompletionsProviderTriggerChangeStrategy = exports.JointCompletionsProviderStrategy = void 0;
var JointCompletionsProviderStrategy;
(function (JointCompletionsProviderStrategy) {
    JointCompletionsProviderStrategy["Regular"] = "regular";
    JointCompletionsProviderStrategy["CursorEndOfLine"] = "cursorEndOfLine";
})(JointCompletionsProviderStrategy || (exports.JointCompletionsProviderStrategy = JointCompletionsProviderStrategy = {}));
var JointCompletionsProviderTriggerChangeStrategy;
(function (JointCompletionsProviderTriggerChangeStrategy) {
    JointCompletionsProviderTriggerChangeStrategy["NoTriggerOnRequestInFlight"] = "noTriggerOnRequestInFlight";
    JointCompletionsProviderTriggerChangeStrategy["NoTriggerOnCompletionsRequestInFlight"] = "noTriggerOnCompletionsRequestInFlight";
    JointCompletionsProviderTriggerChangeStrategy["AlwaysTrigger"] = "alwaysTrigger";
})(JointCompletionsProviderTriggerChangeStrategy || (exports.JointCompletionsProviderTriggerChangeStrategy = JointCompletionsProviderTriggerChangeStrategy = {}));
