// @vscode/chat-lib/dist/src/_internal/util/vs/editor/common/core/text/positionToOffsetImpl.js
"use strict";
//!!! DO NOT modify, this file was COPIED from 'microsoft/vscode'
Object.defineProperty(exports, "__esModule", { value: true });
exports.PositionOffsetTransformer = exports.PositionOffsetTransformerBase = void 0;
exports._setPositionOffsetTransformerDependencies = _setPositionOffsetTransformerDependencies;
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
const arraysFind_1 = require("../../../../base/common/arraysFind");
const offsetRange_1 = require("../ranges/offsetRange");
const position_1 = require("../position");
const range_1 = require("../range");
class PositionOffsetTransformerBase {
    getOffsetRange(range) {
        return new offsetRange_1.OffsetRange(this.getOffset(range.getStartPosition()), this.getOffset(range.getEndPosition()));
    }
    getRange(offsetRange) {
        return range_1.Range.fromPositions(this.getPosition(offsetRange.start), this.getPosition(offsetRange.endExclusive));
    }
    getStringEdit(edit) {
        const edits = edit.replacements.map(e => this.getStringReplacement(e));
        return new Deps.deps.StringEdit(edits);
    }
    getStringReplacement(edit) {
        return new Deps.deps.StringReplacement(this.getOffsetRange(edit.range), edit.text);
    }
    getTextReplacement(edit) {
        return new Deps.deps.TextReplacement(this.getRange(edit.replaceRange), edit.newText);
    }
    getTextEdit(edit) {
        const edits = edit.replacements.map(e => this.getTextReplacement(e));
        return new Deps.deps.TextEdit(edits);
    }
}
exports.PositionOffsetTransformerBase = PositionOffsetTransformerBase;
class Deps {
    static { this._deps = undefined; }
    static get deps() {
        if (!this._deps) {
            throw new Error('Dependencies not set. Call _setDependencies first.');
        }
        return this._deps;
    }
}
/** This is to break circular module dependencies. */
function _setPositionOffsetTransformerDependencies(deps) {
    Deps._deps = deps;
}
class PositionOffsetTransformer extends PositionOffsetTransformerBase {
    constructor(text) {
        super();
        this.text = text;
    }
    get lineStartOffsetByLineIdx() {
        if (!this._lineStartOffsetByLineIdx) {
            this._computeLineOffsets();
        }
        return this._lineStartOffsetByLineIdx;
    }
    get lineEndOffsetByLineIdx() {
        if (!this._lineEndOffsetByLineIdx) {
            this._computeLineOffsets();
        }
        return this._lineEndOffsetByLineIdx;
    }
    _computeLineOffsets() {
        this._lineStartOffsetByLineIdx = [];
        this._lineEndOffsetByLineIdx = [];
        this._lineStartOffsetByLineIdx.push(0);
        for (let i = 0; i < this.text.length; i++) {
            if (this.text.charAt(i) === '\n') {
                this._lineStartOffsetByLineIdx.push(i + 1);
                if (i > 0 && this.text.charAt(i - 1) === '\r') {
                    this._lineEndOffsetByLineIdx.push(i - 1);
                }
                else {
                    this._lineEndOffsetByLineIdx.push(i);
                }
            }
        }
        this._lineEndOffsetByLineIdx.push(this.text.length);
    }
    getOffset(position) {
        const valPos = this._validatePosition(position);
        return this.lineStartOffsetByLineIdx[valPos.lineNumber - 1] + valPos.column - 1;
    }
    _validatePosition(position) {
        if (position.lineNumber < 1) {
            return new position_1.Position(1, 1);
        }
        const lineCount = this.textLength.lineCount + 1;
        if (position.lineNumber > lineCount) {
            const lineLength = this.getLineLength(lineCount);
            return new position_1.Position(lineCount, lineLength + 1);
        }
        if (position.column < 1) {
            return new position_1.Position(position.lineNumber, 1);
        }
        const lineLength = this.getLineLength(position.lineNumber);
        if (position.column - 1 > lineLength) {
            return new position_1.Position(position.lineNumber, lineLength + 1);
        }
        return position;
    }
    getPosition(offset) {
        const idx = (0, arraysFind_1.findLastIdxMonotonous)(this.lineStartOffsetByLineIdx, i => i <= offset);
        const lineNumber = idx + 1;
        const column = offset - this.lineStartOffsetByLineIdx[idx] + 1;
        return new position_1.Position(lineNumber, column);
    }
    getTextLength(offsetRange) {
        return Deps.deps.TextLength.ofRange(this.getRange(offsetRange));
    }
    get textLength() {
        const lineIdx = this.lineStartOffsetByLineIdx.length - 1;
        return new Deps.deps.TextLength(lineIdx, this.text.length - this.lineStartOffsetByLineIdx[lineIdx]);
    }
    getLineLength(lineNumber) {
        return this.lineEndOffsetByLineIdx[lineNumber - 1] - this.lineStartOffsetByLineIdx[lineNumber - 1];
    }
}
exports.PositionOffsetTransformer = PositionOffsetTransformer;
