// @vscode/chat-lib/dist/src/_internal/platform/proxyModels/common/proxyModelsService.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.NullProxyModelsService = exports.IProxyModelsService = void 0;
const services_1 = require("../../../util/common/services");
const event_1 = require("../../../util/vs/base/common/event");
exports.IProxyModelsService = (0, services_1.createServiceIdentifier)('IProxyModelsService');
class NullProxyModelsService {
    constructor() {
        this.onModelListUpdated = event_1.Event.None;
    }
    get models() {
        return undefined;
    }
    get nesModels() {
        return undefined;
    }
    get instantApplyModels() {
        return undefined;
    }
}
exports.NullProxyModelsService = NullProxyModelsService;
